Firebreak: Finesse the Proxy Node pipeline

CharlesIC opened the pull request:
A number of small improvements to the pipeline:
  1) Host metadata statically from apps

  - The changes in this PR will re-publish the existing metadata on the new URLs. If this works, the entity IDs will be changed in a separate PR
  - If this works we can remove a bunch of stuff from the pipeline
  - Added a configurable `MetadataPublishingBundle` that allows to publish a specified resource at a specified URL
  - The metadata resources will be mounted by K8 in the apps' resource directories in their pods

  2) Redirect to different error page per environment
  2) Make Stub Connector prettier and add favicon
  2) Make all Dockerfiles consistent and `wget` less noisy
  2) Make failed package builds not increase the release number until they go green (ripped from #230) 

I based 1) above on the following sources of information:
[Kubernetes docs - hostPath](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath)
[VMC config](https://github.com/alphagov/verify-metadata-controller/blob/f2862119d8b79e9fcbfc937c137cb039afda2747/pkg/controller/metadata/metadata_controller.go)
[Istio VirtualService reference](https://istio.io/docs/reference/config/networking/v1alpha3/virtual-service/#VirtualService)
[Article on routing in VirtualServices](https://rinormaloku.com/istio-practice-routing-virtualservices/)
@chrisfarms

idrop committed:
Firebreak: Failed builds should use the same release number

A failing concourse build at package stage currently requests a new github-release.

Failing package builds will now use the same release until they go green.

Ripped from #230

Co-authored-by: <phillip.miller@digital.cabinet-office.gov.uk>

CharlesIC committed:
EID-1441: Set different error page URL per environment

Author: @CharlesIC

CharlesIC committed:
Firebreak: Improve Stub Connector look & feel

  - Make the EU Trust Mark logo prettier with no pixelation and alpha channel for nicer rendering
  - Add favicon to Stub Connector

Author: @CharlesIC

CharlesIC committed:
Firebreak: Host metadata statically

The Proxy Node and Connector Node metadatas can be hosted on the existing virtual services
This also makes the Connector Node metadata publicly accessible

Author: @CharlesIC

idrop commented on chart/templates/gateway-virtual-service.yaml:
```diff
@@ -13,6 +13,14 @@ spec:
   http:
+    - match:
+      - uri:
+          exact: {{ include "gateway.host" . }}/ServiceMetadata
```
Does the metadata url need to match the entityId of that metadata?
i.e. the entityId of https://test-metadata.london.verify.govsvc.uk/metadata.xml is "https://test-metadata.london.verify.govsvc.uk/metadata.xml". See also https://eidas.minez.nl/EidasNodeC/ConnectorMetadata
I'm not certain its required, but does seem to be a convention.

CharlesIC replied:
Yes it does but we can change that later - this will publish the existing metadata at the new URL in addition to the old one. If this works, we'll change the entity IDs and remove old metadata

idrop replied:
Ok, makes sense

idrop commented on ci/build/release-pipeline.yaml:
```diff
@@ -379,6 +379,7 @@ spec:
             RELEASE_NAME: test
             RELEASE_NAMESPACE: ((namespace-deployer.namespace))
             CLUSTER_PRIVATE_KEY: ((cluster.privateKey))
+            ERROR_PAGE_URL: https://www.integration.signin.service.gov.uk/proxy-node-error
```
👍

CharlesIC committed:
Firebreak: Make all Dockerfiles consistent and wget less noisy

  - Use upper case for Docker commands in all Dockerfiles
  - Force wget progress bar type
  - Serve Stub Connector favicon from `resources/assets/` and add a test

Author: @CharlesIC

CharlesIC committed:
Firebreak: Host metadata statically using a metadata publishing bundle

  - Added `MetadataPublishingBundle` that adds functionality to serve a static metadata file from a predetermined location in the app's resources
  - The config tells the app which resource to publish and at which URL it should be accessible
  - The config params will be provided from K8 config
  - The metadata file will be mounted as a volume in the app's resources by K8

Author: @CharlesIC

CharlesIC committed:
Firebreak: Mount metadata from K8 cluster to Stub Connector and Gateway pods

Added K8 config to mount the Stub Connector and Proxy Node metadata files in the resource directories of their pods.
The apps will the be able to serve their metadata statically.

Author: @CharlesIC

chrisfarms commented on chart/templates/gateway-deployment.yaml:
```diff
@@ -55,3 +67,7 @@ spec:
           value: http://{{ .Release.Name }}-vsp/
         - name: PROXY_NODE_ERROR_PAGE_URI
           value: {{ .Values.gateway.errorPageURL }}
+        - name: PROXY_NODE_METADATA_RESOURCE_FILE_PATH
```
does `PROXY_NODE_METADATA_RESOURCE_FILE_PATH ` and `PROXY_NODE_METADATA_PUBLISH_PATH ` really need to be configurable via the environment?.... looks like they could just be hardcoded or could they have decent defaults so that the amount of config could be reduced.

CharlesIC replied:
Yeah maybe they could - I'm still trying to figure if this will work and will refine this further

chrisfarms commented on chart/templates/stub-connector-deployment.yaml:
```diff
@@ -53,6 +64,10 @@ spec:
           value: "80"
         - name: CONNECTOR_NODE_BASE_URL
           value: https://{{ include "connector.host" . }}
+        - name: CONNECTOR_METADATA_RESOURCE_FILE_PATH
```
Same comment as per `PROXY_NODE_METADATA_RESOURCE_FILE_PATH`

CharlesIC committed:
Firebreak: Show gradle logs in Travis

Author: @CharlesIC

tommrr reviewed (dismissed):
LGTM
